raspberry/main.py

```diff
@@ -0,0 +1,371 @@
+        # lqr.set_Q_and_R_matrix_gains(Q11=LQR_Q11_gain[0], Q22=LQR_Q22_gain[0], R=LQR_R_gain[0])
+
+        if is_drone_flying_close_to_ground[0]:
+            z_hat[0] = current_altitude_snap_shot_mts[0] if temp == - \
+                1 else temp/1000
+            print(
+                f"Cannot activate Altitude hold. Drone is flying close to the ground at {last_valid_altitude_measurement_mts[0]/1000} mts < {min_altitude_to_activate_AltiHold_mts[0]} mts.")
+            if is_KF_ran_atleast_once[0]:
+                kf.reset()
+        else:
+            x_est = kf.update(Tref_t[0], euler[1], euler[2],
+                           np.nan if temp == -1 else temp/1000)
+            is_KF_ran_atleast_once[0] = True
+            z_hat[0] = x_est[0][0]
+            v_hat[0] = x_est[1][0]
+            alpha_hat[0] = x_est[2][0]
+            beta_hat[0] = x_est[3][0]
+
+            KF_data[0] = z_hat[0]
+            KF_data[1] = v_hat[0]
+            KF_data[2] = alpha_hat[0]
+            KF_data[3] = beta_hat[0]
+
+            if use_altitude_hold[0]:
+                if not is_current_altitude_snap_shot_taken[0]:
+                    current_altitude_snap_shot_mts[0] = temp/1000
+                    var_e_RC_mid_percentage[0] = rc.trimmer_VRE_percentage()
+                    is_current_altitude_snap_shot_taken[0] = True
+            else:
+                is_current_altitude_snap_shot_taken[0] = False
+
+        if use_altitude_hold[0] and not is_drone_flying_close_to_ground[0]:
+            throttle_ref_from_LQR[0] = lqr.control_action(np.array([[z_hat[0]], [v_hat[0]]]),
+                                                          alpha_t=alpha_hat[0],
+                                                          beta_t=beta_hat[0],
+                                                          reference_altitude_mts=altitude_ref_mts[0],
+                                                          recalculate_dynamics=True,
+                                                          pitch_rad=euler[1],
+                                                          roll_rad=euler[2],
+                                                          k11=-LQR_Q11_gain[0],
+                                                          k12=-LQR_Q22_gain[0])
+            throttle_ref_from_LQR[0] = np.array(
+                [[max(1000, min(throttle_ref_from_LQR[0][0][0]*900, 600) + 1000)]])
+            Tref_t[0] = (throttle_ref_from_LQR[0][0][0] - 1000)/900
+            print(
+                f"alphan hat: {alpha_hat[0]} beta hat: {beta_hat[0]} k11: {-LQR_Q11_gain[0]} k12: {-LQR_Q22_gain[0]} Tref_LQR: {Tref_t[0]} alt_hat: {z_hat[0]} alt_ref: {altitude_ref_mts[0]}")
+
+    if enable_caching[0]:
+        time_before_thread_starts[0] = time_ns()
+
+    # schedule the necessary functions
+    scheduler.schedule("process_radio_data",
+                       process_radio_data,
+                       function_call_frequency=50,
```

Comment:
should this be `function_call_frequency=sampling_frequency`? (instead of hard-coding the value 50)

Reply:
No, we might want to run the other process at different rates than `LQR and KF`. Although for now, they happen to have the same `function_call_frequency`, they should not be set to `sampling_frequency`. Because in future we will have to change `sampling_frequency` to a different value based on what the new **better quality** sensors support, which could be something different than `50 Hz`. But we will always want to run communication tasks between `ESP` and `R-Pi` at `50 Hz` and hence hard-coded values.